### PR TITLE
2021.3 Shader Support

### PR DESCRIPTION
**Pull request by GZHYBFHHJ**
Related issue: #1040

### CODE REVIEW

**Comment by ds5678** [comment]
You have a lot of indentation changes, which make this significantly more difficult to review. It will not make it into tomorrow's release and might not make it into next week's release, depending on the time I have available.

Thank you for your time spent working on this. I appreciate your contribution.

**Comment by GZHYBFHHJ** [comment]
difftastic should handle indentation better, but it doesn't seem to work on Github Desktop.
I may try to re-commit using the CLI later.

**Comment by ds5678** [comment]
It doesn't matter if another diff software handles indentation better. I can only leave review comments on GitHub.

If the format is substantially different, I would prefer you have separate methods or classes instead of if checking everything.

**Comment by Futuremappermydud** [comment]
image
Maybe worth a mention, this does not work for me.
The game is beat saber. Unity version is `2021.3.16f1`

**Comment by GZHYBFHHJ** [comment]
This bug is not related to this pull request. It also appears in games made with Unity 2021.2.
It is because of `m_CommonParameters` (in 2020.3+)

**Comment by Futuremappermydud** [comment]
> Duplication of #726 

Thanks gonna try their fix out. Just though I'd throw it in here since you seem to be working on it right now

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs [hunk 1]
```diff
@@ -7,18 +9,56 @@ public static class SerializedProgramExtensions
...
-					break;
-				}
+					if (program.SubPrograms[i].ShaderHardwareTier <= tier)
+					{
+						break;
+					}
 
-				tierCount++;
+					tierCount++;
+				}
```
Comment (ds5678) [inline Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:27]: Why was this changed? `ISerializedProgram.SubPrograms` and `ISerializedSubProgram.ShaderHardwareTier` are present on all Unity versions.

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs [hunk 2]
```diff
@@ -7,18 +9,56 @@ public static class SerializedProgramExtensions
...
+			{
+				for (int i = 0; i < program.PlayerSubPrograms.Count; i++)
+				{
+					if (program.PlayerSubPrograms[i].Count > 0)
+					{
+						return program.PlayerSubPrograms[i];
+					}
+				}
+			}
+			return new AssetList<SerializedPlayerSubProgram>();
```
Comment (ds5678) [inline Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:45]: Change the return type to `IReadOnlyList<SerializedPlayerSubProgram>` and this line to `return [];`

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs [hunk 3]
```diff
@@ -7,18 +9,56 @@ public static class SerializedProgramExtensions
...
+			{
+				for (int i = 0; i < program.ParameterBlobIndices.Count; i++)
+				{
+					if (program.ParameterBlobIndices[i].Count > 0)
+					{
+						return program.ParameterBlobIndices[i];
+					}
+				}
+			}
+			return new AssetList<uint>();
```
Comment (ds5678) [inline Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:60]: As above

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs (hunk 2 above)
Comment (ds5678) [reply Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:45]: Actually, throwing an exception here might be better if the result should never have a count of 0. Otherwise, we'll just get an index out of range exception later.

File: Source/AssetRipper.Export.UnityProjects/Shaders/USCShaderExporter.cs [hunk 4]
```diff
@@ -562,10 +585,19 @@ private static List<ShaderSubProgram> GetSubPrograms(IShader shader, ShaderSubPr
 		{
 			List<ShaderSubProgram> matchingPrograms = new();
 			ShaderSubProgram? fallbackProgram = null;
-			for (int i = 0; i < program.SubPrograms.Count; i++)
+
+			AssetList<SerializedPlayerSubProgram> playerSubPrograms = program.GetPlayerSubPrograms();
+			for (int i = 0; program.Has_PlayerSubPrograms() ? i < playerSubPrograms.Count : i < program.SubPrograms.Count; i++)
```
Comment (ds5678) [inline Source/AssetRipper.Export.UnityProjects/Shaders/USCShaderExporter.cs:590]: Store the count in a local.

File: Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs [hunk 5]
```diff
@@ -33,95 +33,19 @@ public void Read(AssetCollection shaderCollection, byte[] compressedBlob, AssetL
...
-			{
-				ShaderSubProgramEntry entry = Entries[i];
-				if (entry.Segment == segment)
-				{
-					writer.BaseStream.Position = entry.Offset;
-					SubPrograms[i].Write(writer);
-				}
-			}
-		}
+		public void Write(AssetCollection shaderCollection, MemoryStream memStream, out uint[] offsets, out uint[] compressedLengths, out uint[] decompressedLengths) => throw new NotImplementedException();
```
Comment (ds5678) [inline Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs:48]: Just delete it.

File: Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs [hunk 6]
```diff
@@ -33,95 +33,19 @@ public void Read(AssetCollection shaderCollection, byte[] compressedBlob, AssetL
...
-			using MemoryStream blobMem = new MemoryStream(decompressedBuffer);
+			using MemoryStream blobMem = new MemoryStream(m_decompressedBlob);
 			using AssetReader blobReader = new AssetReader(blobMem, shaderCollection);
 			if (segment == 0)
 			{
 				Entries = ReadAssetArray(blobReader);
-				SubPrograms = CreateAndInitializeArray<ShaderSubProgram>(Entries.Length);
+				m_cachedSubPrograms.Clear();
 			}
-			ReadSegment(blobReader, segment);
```
Comment (ds5678) [inline Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs:46]: Is there a reason these can't be read all at once?

File: Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgram.cs [hunk 7]
```diff
@@ -302,7 +321,7 @@ public void Read(AssetReader reader)
 			}
 		}
 
-		public void Write(AssetWriter writer) => throw new NotImplementedException();
+		public void Write(AssetWriter writer, bool writePlayerSubProgram = true, bool writeParameter = true) => throw new NotImplementedException();
```
Comment (ds5678) [inline Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgram.cs:324]: Delete it

File: Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgram.cs [hunk 8]
```diff
@@ -60,7 +60,7 @@ private static int GetExpectedProgramVersion(UnityVersion version)
 			};
 		}
 
-		public void Read(AssetReader reader)
+		public void Read(AssetReader reader, bool readProgramData = true, bool readParams = true)
```
Comment (ds5678) [inline Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgram.cs:63]: These additional parameters are not intuitive and should be documented.

File: Source/AssetRipper.Export.Modules.Shader/IO/SerializedExtensions.cs [hunk 9]
```diff
@@ -576,7 +616,7 @@ public static void Export(this ShaderSubProgramBlob _this, ShaderWriter writer,
 				}
 
 				// we don't know shader type so pass vertex
-				_this.SubPrograms[subIndex].Export(writer, ShaderType.Vertex);
+				_this.GetSubProgram(writer.Shader.Collection, (uint)subIndex).Export(writer, ShaderType.Vertex);
```
Comment (ds5678) [inline Source/AssetRipper.Export.Modules.Shader/IO/SerializedExtensions.cs:619]: I think we should just store the collection on the blob instead of passing it as a parameter.

**Review by ds5678 — CHANGES_REQUESTED** [review]
Hello, I found some time to review your pr. I appreciate all the work you've put into it.

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs (hunk 1 above)
Comment (GZHYBFHHJ) [reply Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:27]: `ISerializedProgram.SubPrograms` is always empty if `ISerializedProgram.PlayerSubProgram` is not null

File: Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs (hunk 6 above)
Comment (GZHYBFHHJ) [reply Source/AssetRipper.Export.Modules.Shader/ShaderBlob/ShaderSubProgramBlob.cs:46]: It is not possible to directly determine whether an entry only contains ProgramData or Parameters here.

File: Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs (hunk 1 above)
Comment (ds5678) [reply Source/AssetRipper.SourceGenerated.Extensions/SerializedProgramExtensions.cs:27]: For now, let's just account for the situation where it's empty.
